### PR TITLE
Schrödinger's Cat

### LINKED ISSUE #22054: Optional dependent is not null when sharing a required property
```C#
namespace efcore.ownedentities
{
    using Microsoft.EntityFrameworkCore;
    using System;
    using System.Linq;

    class Program
    {
        static void Main(string[] args)
        {
            var id = Guid.NewGuid();

            using (var db = new Ctx())
            {
                db.Database.EnsureDeleted();
                db.Database.EnsureCreated();

                var p = new User(id){ Mobile = null };
                db.Entities.Add(p);
                db.SaveChanges();
            }

            using (var db = new Ctx())
            {
                var entity = db.Find<User>(id);
                Assert.Null(entity.Mobile);
            }
        }

        public class User // parent entity
        {
            private Guid _id;

            public User(Guid id)
            {
                _id = id;
            }

            public Guid Id => _id;

            public MobileNumber Mobile { get; set; }

            public byte[] RowVersion { get; set; }
        }

        public class MobileNumber // value object
        {
            public string Number { get; set; }
        }

        public class Ctx : DbContext
        {
            public DbSet<User> Entities { get; set; }

            protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
            {
                optionsBuilder.UseMySql("Server = localhost; Database = efownedentities; User = root; Password = admin; ");
            }

            protected override void OnModelCreating(ModelBuilder mb)
            {
                var parentEntityBuilder = mb.Entity<User>();

                parentEntityBuilder.HasKey(x => x.Id);

                parentEntityBuilder.Property(x => x.Id)
                    .HasField("_id")
                    .UsePropertyAccessMode(PropertyAccessMode.Field)
                    .IsRequired();

                parentEntityBuilder.Property(x => x.Id).ValueGeneratedNever();

                parentEntityBuilder
                    .Property<byte[]>(x => x.RowVersion)
                    .HasColumnType("TIMESTAMP")
                    .IsRowVersion()
                    .IsRequired()
                    .HasColumnName("RowVersion");

                parentEntityBuilder
                    .OwnsOne(o => o.Mobile, cfg =>
                    {
                        cfg.Property(x => x.Number)
                            .HasMaxLength(30);

                        cfg.Property<byte[]>("RowVersion")
                            .IsRowVersion()
                            .IsRequired()
                            .HasColumnType("TIMESTAMP")
                            .HasColumnName("RowVersion");
                    });
            }
        }
    }
}
```

### PR DESCRIPTION
**Pull request by smitpatel**
**Query: New set of conditions to materialize optional dependents**

Let A be the set of properties which are not sharing column with any principal.
B be the subset of A where property is required.
C be subset of A where property is optional. (hence A = B + C)

The necessary condition for materialization without anything else be C1 - all required properties (excluding PK) must have a value (with or without sharing).
- If A is empty, then C1 will be sufficient condition as it will be required dependent.
- If B is non-empty, then C1 will cover those properties and be sufficient condition.
- If B is empty and then C is non-empty, then materialize if there is at least 1 non-null value in C along with C1 being true.
- Otherwise null.

If there are no required properties other than PK (essentially C1 does not exist). B is by definition empty.
- If A is empty then always materialize.
- If A is non-empty implies C is non-empty, then materialize if there is at least 1 non-null value in C.

Resolves #22054

### CODE REVIEW

File: src/EFCore.Relational/Query/RelationalEntityShaperExpression.cs [hunk 1]
```diff
@@ -103,37 +104,43 @@ protected override LambdaExpression GenerateMaterializationCondition(IEntityType
...
+                            .Select(p => NotEqual(
+                                valueBufferParameter.CreateValueBufferReadValueExpression(typeof(object), p.GetIndex(), p),
+                                Constant(null)))
+                            .Aggregate((a, b) => AndAlso(a, b));
                     }
-                    else
+
+                    var allNonSharedProperties = GetNonSharedProperties(table, entityType);
+                    if (allNonSharedProperties.Count != 0
+                        && allNonSharedProperties.Count(p => !p.IsNullable) == 0)
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Query/RelationalEntityShaperExpression.cs:127]: Can use Any

Comment (smitpatel) [reply src/EFCore.Relational/Query/RelationalEntityShaperExpression.cs:127]:
```suggestion
                        && allNonSharedProperties.Any(p => !p.IsNullable))
```

File: src/EFCore.Relational/Query/RelationalEntityShaperExpression.cs [hunk 2]
```diff
@@ -169,5 +176,39 @@ public override EntityShaperExpression Update(Expression valueBufferExpression)
...
+            foreach (var property in entityType.GetProperties())
+            {
+                if (property.IsPrimaryKey())
+                {
+                    continue;
+                }
+
+                var propertyMappings = table.FindColumn(property).PropertyMappings;
+                if (propertyMappings.Count() > 1
+                    && propertyMappings.Any(pm => principalEntityTypes.Any(et => !pm.Property.DeclaringEntityType.IsAssignableFrom(et))))
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Query/RelationalEntityShaperExpression.cs:194]:
```suggestion
                    && propertyMappings.Any(pm => principalEntityTypes.Contains(pm.TableMapping.EntityType)))
```

File: src/EFCore.Relational/Query/RelationalEntityShaperExpression.cs (hunk 1 above)
Comment (AndriySvyryd) [reply src/EFCore.Relational/Query/RelationalEntityShaperExpression.cs:127]: It should be
```C#
&& !allNonSharedProperties.Any(p => !p.IsNullable))
```

**Pull request merged**